verilog/analysis/checkers/instance_shadow_rule.cc

```diff
@@ -0,0 +1,135 @@
+const char InstanceShadowRule::kTopic[] = "mark-shadowed-instances";
+const char InstanceShadowRule::kMessage[] =
+    "Instance shadows the already declared variable";
+
+std::string InstanceShadowRule::GetDescription(
+    DescriptionType description_type) {
+  return absl::StrCat(
+      "Checks that any defined variable does not shadow other variable in the "
+      "same scope",
+      GetStyleGuideCitation(kTopic), ".");
+}
+
+static const Matcher& InstanceShadowMatcher() {
+  static const Matcher matcher(SymbolIdentifierLeaf());
+  return matcher;
+}
+static bool isInAllowedNode(const SyntaxTreeContext& ctx) {
+  return ctx.IsInside(NodeEnum::kSeqBlock) ||
+         ctx.IsInside(NodeEnum::kGenvarDeclaration) ||
+         ctx.IsInside(NodeEnum::kReference);
+}
+
+void InstanceShadowRule::HandleSymbol(const verible::Symbol& symbol,
+                                      const SyntaxTreeContext& context) {
+  verible::matcher::BoundSymbolManager manager;
+  bool omit_node = false;
+  if (!InstanceShadowMatcher().Matches(symbol, &manager)) {
+    return;
+  }
+  const auto& labels = FindAllSymbolIdentifierLeafs(symbol);
+  if (labels.empty()) return;
+  // if the considered symbol name is not a declaration
+  if (context.IsInside(NodeEnum::kReference)) return;
+
+  const auto& rcontext = reversed_view(context);
+  const auto& rdirectParent = *std::next(rcontext.begin());
+  // we are looking for the potential labels that might overlap the considered
+  // declaration. We are searching all the labels within the visible scope
+  // until we find the node or we reach the top of the scope
+  for (auto rc = rcontext.begin(); rc != rcontext.end(); rc++) {
+    const auto& node = *rc;
+
+    for (const verible::SymbolPtr& child : node->children()) {
+      if (child == nullptr) {
+        continue;
+      }
+      const auto& overlappingLabels = FindAllSymbolIdentifierLeafs(*child);
+      for (const verible::TreeSearchMatch& omatch : overlappingLabels) {
+        const auto& overlappingLabel = SymbolCastToLeaf(*omatch.match);
+        // variable in different scopes or this is not a
+        // vulnerable declaration
+        if (isInAllowedNode(omatch.context)) {
+          omit_node = true;
+          break;
+        }
+        const auto& label = SymbolCastToLeaf(*labels[0].match);
+        // if found label has the same adress as considered label
+        // we found the same node so we don't
+        // want to look further
+        if (omatch.match == labels[0].match) {
+          omit_node = true;
+          break;
+        }
+        // if considered label is the last node
+        // this is the ending node with label
+        if (rdirectParent->children().back().get() ==
+            node->children().back().get()) {
+          return;
+        }
+
+        if (overlappingLabel.get().text() == label.get().text()) {
+          violations_.insert(LintViolation(symbol, kMessage, context));
```

Comment:
I think we should make sure to provide some actionable information. In particular in larger designs it might be hard to find what has been shadowed. I am thinking of an error message of the form
```
systemverilog-file.sv:12:15: Symbol `foo` is shadowing symbol `foo` defined at systemverilog-file:3:7:
```

Reply:
I have added the new mechanism for injecting the location of the tokens into the vulnerability description.
What do you think about this solution?
This allows us to put some additional tokens and mark the location in the text,
then substitute the markers `\@` with tokens location.